File: 0894_All_Possible_Full_Binary_Trees/solution.java
package cn.huang.leetcode;

import cn.huang.leetcode.common.TreeNode;

import java.util.ArrayList;
import java.util.List;

/*
894. All Possible Full Binary Trees
A full binary tree is a binary tree where each node has exactly 0 or 2 children.

Return a list of all possible full binary trees with N nodes.  Each element of the answer is the root node of one possible tree.

Each node of each tree in the answer must have node.val = 0.

You may return the final list of trees in any order.



Example 1:

Input: 7
Output: [[0,0,0,null,null,0,0,null,null,0,0],[0,0,0,null,null,0,0,0,0],[0,0,0,0,0,0,0],[0,0,0,0,0,null,null,null,null,0,0],[0,0,0,0,0,null,null,0,0]]
 */
public class LC_0894_AllPossibleFullBinaryTrees {
    public List<TreeNode> allPossibleFBT(int N) {
        List<TreeNode> ans = new ArrayList<>();
        if (N % 2 == 0)  //full binary tree each node has exactly 0 or 2 children
            return ans;
        if (N == 1) {
            ans.add(new TreeNode(0));
            return ans;
        }
        for (int i = 1; i < N; i += 2) {
            for (TreeNode l : allPossibleFBT(i))
                for (TreeNode r : allPossibleFBT(N - i - 1)) {
                    TreeNode root = new TreeNode(0);
                    root.left = l;
                    root.right = r;
                    ans.add(root);
                }
        }
        return ans;
    }
}

File: 0894_All_Possible_Full_Binary_Trees/solution.cpp
/*
894. All Possible Full Binary Trees
https://leetcode.com/problems/all-possible-full-binary-trees/
 */
#include <vector>
using namespace std;

#include "common/TreeNode.h"

/*
https://zxi.mytechroad.com/blog/tree/leetcode-894-all-possible-full-binary-trees/
 */
class Solution
{
public:
  vector<TreeNode *> allPossibleFBT(int N)
  {
    if (N % 2 == 0)
      return {};
    if (N == 1)
      return {new TreeNode(0)};
    vector<TreeNode *> ans;
    for (int i = 1; i < N; i += 2)
    {
      for (const auto &l : allPossibleFBT(i))
        for (const auto &r : allPossibleFBT(N - i - 1))
        {
          auto root = new TreeNode(0);
          root->left = l;
          root->right = r;
          ans.push_back(root);
        }
    }
    return ans;
  }
};